fluent-bundle/src/bundle.ts

```diff
@@ -1,12 +1,33 @@
-import {FluentNone} from "./types.js";
-import {resolveComplexPattern} from "./resolver.js";
-import Scope from "./scope.js";
+import { resolveComplexPattern } from "./resolver.js";
+import { Scope } from "./scope.js";
+import { FluentError } from "./error.js";
+import { FluentResource } from "./resource.js";
+import { FluentNone, FluentType } from "./types.js";
+import { RuntimeMessage, RuntimeTerm, RuntimePattern } from "./ast.js";
+
+export type CustomFunction = (
+  positional: Array<FluentType>,
+  named: Record<string, FluentType>
+) => FluentType;
+
+export type CustomTransform = (text: string) => string;
+
+export type FluentArgument = string | number | Date | FluentType;
 
 /**
  * Message bundles are single-language stores of translation resources. They are
  * responsible for formatting message values and attributes to strings.
  */
-export default class FluentBundle {
+export class FluentBundle {
+  public locales: Array<string>;
+
+  public _terms: Map<string, RuntimeTerm> = new Map();
+  public _messages: Map<string, RuntimeMessage> = new Map();
+  public _functions: Record<string, CustomFunction>;
+  public _useIsolating: boolean;
+  public _transform: CustomTransform;
+  public _intls: WeakMap<any, any> = new WeakMap();
```

Comment:
What's a good way to type the `Intl` formatters here? A union of all existing ones sounds like something which eventually becomes out of date.

Reply:
I'm afraid I got nothing. :(

I've tried a couple of ideas. The stricter ones got stuck being incompatible with some other types. Other, more relaxed types seem marginally better than `any` and might not be worth it.

I can tell that the type `NonNullable<Parameters<typeof Intl.DateTimeFormat>[1]>` will give `Intl.DateTimeFormatOptions` (works for `Intl.NumberFormat` too), but I couldn't figure out how to get TypeScript to infer `typeof value` from usage. And that's about the only potentially useful thing that came out of my fiddling with the typings here and in the `Scope` class.

Reply:
Thanks. I went with `WeakMap<object, Record<string, object>>` for now which is only marginally better than `any`.